airbyte-cdk/bulk/toolkits/load-object-storage/src/main/kotlin/io/airbyte/cdk/load/file/object_storage/ObjectStorageFormattingWriter.kt

```diff
@@ -0,0 +1,127 @@
+
+import io.airbyte.cdk.load.command.DestinationStream
+import io.airbyte.cdk.load.command.object_storage.AvroFormatConfiguration
+import io.airbyte.cdk.load.command.object_storage.CSVFormatConfiguration
+import io.airbyte.cdk.load.command.object_storage.JsonFormatConfiguration
+import io.airbyte.cdk.load.command.object_storage.ObjectStorageFormatConfigurationProvider
+import io.airbyte.cdk.load.command.object_storage.ParquetFormatConfiguration
+import io.airbyte.cdk.load.data.DestinationRecordToAirbyteValueWithMeta
+import io.airbyte.cdk.load.data.avro.toAvroRecord
+import io.airbyte.cdk.load.data.avro.toAvroSchema
+import io.airbyte.cdk.load.data.csv.toCsvRecord
+import io.airbyte.cdk.load.data.json.toJson
+import io.airbyte.cdk.load.file.avro.toAvroWriter
+import io.airbyte.cdk.load.file.csv.toCsvPrinterWithHeader
+import io.airbyte.cdk.load.file.parquet.toParquetWriter
+import io.airbyte.cdk.load.message.DestinationRecord
+import io.airbyte.cdk.load.util.serializeToString
+import io.airbyte.cdk.load.util.write
+import io.micronaut.context.annotation.Secondary
+import jakarta.inject.Singleton
+import java.io.Closeable
+import java.io.OutputStream
+
+interface ObjectStorageFormattingWriter : Closeable {
+    fun accept(record: DestinationRecord)
+}
+
+@Singleton
+@Secondary
+class ObjectStorageFormattingWriterFactory(
+    private val recordDecorator: DestinationRecordToAirbyteValueWithMeta,
+    private val formatConfigProvider: ObjectStorageFormatConfigurationProvider,
+) {
+    fun create(
+        stream: DestinationStream,
+        outputStream: OutputStream
+    ): ObjectStorageFormattingWriter {
+        return when (formatConfigProvider.objectStorageFormatConfiguration) {
+            is JsonFormatConfiguration -> JsonFormattingWriter(outputStream, recordDecorator)
+            is AvroFormatConfiguration ->
+                AvroFormattingWriter(
+                    stream,
+                    outputStream,
+                    formatConfigProvider.objectStorageFormatConfiguration
+                        as AvroFormatConfiguration,
+                    recordDecorator
+                )
+            is ParquetFormatConfiguration ->
+                ParquetFormattingWriter(
+                    stream,
+                    outputStream,
+                    formatConfigProvider.objectStorageFormatConfiguration
+                        as ParquetFormatConfiguration,
+                    recordDecorator
+                )
+            is CSVFormatConfiguration -> CSVFormattingWriter(stream, outputStream, recordDecorator)
+        }
+    }
+}
+
+class JsonFormattingWriter(
+    private val outputStream: OutputStream,
+    private val recordDecorator: DestinationRecordToAirbyteValueWithMeta
+) : ObjectStorageFormattingWriter {
+    override fun accept(record: DestinationRecord) {
+        outputStream.write(recordDecorator.decorate(record).toJson().serializeToString())
+    }
+
+    override fun close() {
+        // Do nothing
+    }
```

Comment:
coderabbit is wrong (I think), but it's probably best to add a comment explaining that these writers explicitly should (must?) not close the outputstream (b/c S3MultipartUpload does an`UploadStream().use {}`)

Reply:
The contract on `Closeable::close` is that it conflates. Multiple calls should be ignored. Technically something like this

```
inputStream.use { is ->
myWriter(is).use { mw ->
  mw.stuff()
}
is.stuff()
}
```

will break for writers like avro/parquet that close the underlying stream, but there's no risk of that given the way we use it here. (`streamingUpload` behaves like delegation is *possible*, but not *required.* I'll add comments to that effect.)

avro/parquet close because they are *file* writers: they're writing headers and closing data. The Json writer is just writing lines, so maybe technically shouldn't close? Though maybe it should close just to match the behavior of the other writers?

Reply:
I went ahead and made it close for parity. I also added comments to the `streamingUpload` interface.